src/lib/structures/Command.ts

```diff
@@ -96,8 +98,65 @@ export abstract class Command<T = Args> extends AliasPiece {
 			strategy: this.strategy
 		};
 	}
+
+	protected resolveConstructorPreConditions(options: CommandOptions): readonly PreconditionEntryResolvable[] {
+		const preconditions = (options.preconditions ??= []) as PreconditionEntryResolvable[];
+		if (options.nsfw) preconditions.push('NSFW');
+
+		const runIn = this.resolveConstructorPreConditionsRunType(options.runIn);
+		if (runIn !== null) preconditions.push(runIn);
+
+		const coolDownBucket = options.cooldownBucket ?? 1;
+		if (coolDownBucket && options.cooldownDuration) {
+			preconditions.push({ name: 'Cooldown', context: { bucket: coolDownBucket, cooldown: options.cooldownDuration } });
+		}
+
+		return preconditions;
+	}
+
+	private resolveConstructorPreConditionsRunType(runIn: CommandOptions['runIn']): CommandOptionsRunPreConditions[] | null {
+		if (isNullish(runIn)) return null;
+		if (typeof runIn === 'string') {
+			switch (runIn) {
+				case 'dm':
+					return ['DMOnly'];
+				case 'text':
+					return ['TextOnly'];
+				case 'news':
+					return ['NewsOnly'];
+				case 'guild':
+					return ['GuildOnly'];
+				default:
+					return null;
+			}
+		}
+
+		// If there's no channel it can run on, throw an error:
+		if (runIn.length === 0) {
```

Comment:
shouldn't this be checked first?

Reply:
That'd be either `null.length` or `undefined.length`, which would throw an error.
And any earlier would be `string.length`, which doesn't match the types and defaults to `null`.
This is the first check as we know as "runIn is an array" thanks to the type shortening of the previous two checks.